blog/2021-05-29-using-rust.md

```diff
@@ -0,0 +1,74 @@
+
+## Lifecycle of an image
+
+Before we dive into the infrastructure, it would help to take a look at what an image goes through once it enters our system.
+
+- The image is assigned a UUID.
+- The image is uploaded to a `s3` bucket.
+- The image is then downloaded, preprocessed, run through our models, and a food tip is generated for the image.
+- The food tip, inference results and the image hash form the final response.
+
+## Stack decisions: Rust and haskell
+
+Haskell is used for the client facing API, and it's for other ad-hoc tasks such as reading/writing to a database among others.
+
+Rust is used for image preprocessing, model inference and sending the results back. Sending of results was done differently in the two approaches outlined below. Rust was chosen for it's high efficiency, small executable footprint, and absence of a garbage collector. It also had strong type system, speed & relatively actively maintained Tensorflow (client) library.
+
+# Introduction and pitfalls of the existing architecture
+
+![](/images/v2-arch-diagram.png)
+The internal organization of the rust service in this architecture is outlined above.
+There were 3 main parts, all running concurrently on 3 separate tokio runtimes - namely polling sqs, preprocessing and inferring from the images, and cleanup tasks.
+
+The external processes related to this architecture are outlined below.
+![](/images/v2-external-arch-diagram.png)
+
+The main gripe we had was in the `S3` to `SQS` upload event notification. In our experiments, it was [very slow](https://github.com/Holmusk/aws_benchmarks/tree/master/s3_event_to_sqs), and we aim for the service to have a very low latency, with the goal being that every image that comes into the system should be scored/rated in under 1 second. Because of the way the system was designed, this meant that we'd need a pretty big makeover on the rust side, and some tweaks on the haskell side. This is also mentioned in the [AWS docs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html), where they point that `Typically, event notifications are delivered in seconds but can sometimes take a minute or longer`.
+
+# The new architecture
+
+Like mentioned before, the main reason for redesigning the architecture was to avoid the `S3` to `SQS` upload event notification as speed was of high importance. In the process, we found out that we actually simplified it, by removing unnecessary moving parts.
+
+Internally, the rust service now also has a webserver. The client facing API proxies the requests it receives to the rust webservers via a load balancer. In this version, we completely eliminate the use of SQS.
+
+## Rust webserver internals
+
+![](/images/v3-arch-diagram.png)
+
+We use a [`warp`](https://github.com/seanmonstar/warp) for the webserver in rust.
```

Comment:
```suggestion
We have chosen to use [`warp`](https://github.com/seanmonstar/warp) for the web server implementation in Rust.
```

Add a one-liner explaining our decision of `warp` and why it fits our use-case. 

Later on in the post you can also add any cons/limitations that we faced down the road after making the decision to use `warp` (I'm personally curious and I need to know this for future maintenance purposes 😅 )

Reply:
Cool! We actually didn't really face any limitations with warp tbh. It had support for everything we needed and then some. So not sure what i would add onto it. 
Was actually really fun to use. Excellent tokio interop and the `Filter` system, which uses macros to define the routes and the arguments were really elegant and straightforward :D